### PR TITLE
spirv-fuzz: add tests for full coverage of TransformationAccessChain

### LINKED ISSUE #4286: Achieve full test coverage of TransformationAccessChain
The regular spirv-fuzz tests miss some statements in this transformation class, and it would be good to cover them.

### PR DESCRIPTION
**Pull request by sliu-UIUC**
This PR will fix #4286. 

Update: 
Fixed redundant code. Improved comments and styles. Achieved 100% coverage.

### CODE REVIEW

**Review by afd — CHANGES_REQUESTED** [review]
Thanks - this looks good! A few comments about your comments :)

I agree with your analysis that the if condition can never be true.

I suggest we do a few changes to make the code clearer.

First, since `GetIndexValue` is only called on structs, let's change its name to `GetStructIndexValue`

Second, let's change its comment (in the header file) to reflect this, e.g. remove:

```
  // - the object being indexed is not a composite type
```

and start the whole comment with:

```
  // Requires that |object_type_id| is a struct type
```

Add an assertion at the top of the function to check that this is true:

```
assert(ir_context->get_def_use_mgr()->GetDef(object_type_id) == SpvOpTypeStruct && "Precondition: the type must be a struct type.");
```

And finally, change the redundant if condition into an assertion:

```
assert(spvOpcodeIsConstant(index_instruction->opcode()) && "A non-constant index should already have been rejected.");
```

In the process of investigating this I wrote the test below. Even though it doesn't increase coverage I think we may as well include it. So can you please add it to your PR?

Thanks!

```
TEST(TransformationAccessChainTest, StructIndexMustBeConstant) {
  std::string shader = R"(
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %4 "main"
               OpExecutionMode %4 OriginUpperLeft
               OpSource ESSL 320
          %2 = OpTypeVoid
          %3 = OpTypeFunction %2
          %6 = OpTypeInt 32 1
         %20 = OpUndef %6
          %7 = OpTypeStruct %6 %6
          %8 = OpTypePointer Function %7
         %10 = OpConstant %6 0
         %11 = OpConstant %6 2
         %12 = OpTypePointer Function %6
          %4 = OpFunction %2 None %3
          %5 = OpLabel
          %9 = OpVariable %8 Function
               OpReturn
               OpFunctionEnd
  )";

  const auto env = SPV_ENV_UNIVERSAL_1_4;
  const auto consumer = nullptr;
  const auto context = BuildModule(env, consumer, shader, kFuzzAssembleOption);
  spvtools::ValidatorOptions validator_options;
  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
                                               kConsoleMessageConsumer));
  TransformationContext transformation_context(
      MakeUnique<FactManager>(context.get()), validator_options);
  // Bad: %9 is a pointer to a struct, but %20 is not a constant.
  ASSERT_FALSE(TransformationAccessChain(
                   100, 9, {20}, MakeInstructionDescriptor(9, SpvOpReturn, 0))
                   .IsApplicable(context.get(), transformation_context));
}
```

File: test/fuzz/transformation_access_chain_test.cpp [hunk 1]
```diff
@@ -127,6 +127,29 @@ TEST(TransformationAccessChainTest, BasicTest) {
   transformation_context.GetFactManager()->AddFactValueOfPointeeIsIrrelevant(
       54);
 
+  // Additional tests for full coverage.
```
[the remark this answers is not in the document]
Comment (sliu-UIUC) [reply test/fuzz/transformation_access_chain_test.cpp:130]: Done.

File: test/fuzz/transformation_access_chain_test.cpp [hunk 2]
```diff
@@ -127,6 +127,29 @@ TEST(TransformationAccessChainTest, BasicTest) {
...
+  // Index type is not a 32-bit integer.
```
[the remark this answers is not in the document]
Comment (sliu-UIUC) [reply test/fuzz/transformation_access_chain_test.cpp:131]: Changed.

File: test/fuzz/transformation_access_chain_test.cpp [hunk 3]
```diff
@@ -304,6 +327,19 @@ TEST(TransformationAccessChainTest, BasicTest) {
     ASSERT_FALSE(
         transformation_context.GetFactManager()->PointeeValueIsIrrelevant(107));
   }
+  {
+    // Additional test for coveraging access chain pointee irrelevance.
```
[the remark this answers is not in the document]
Comment (sliu-UIUC) [reply test/fuzz/transformation_access_chain_test.cpp:331]: Fixed.

File: test/fuzz/transformation_access_chain_test.cpp [hunk 4]
```diff
@@ -532,6 +595,7 @@ TEST(TransformationAccessChainTest, ClampingVariables) {
                                                kConsoleMessageConsumer));
   TransformationContext transformation_context(
       MakeUnique<FactManager>(context.get()), validator_options);
+
```
Comment (afd) [inline test/fuzz/transformation_access_chain_test.cpp:598]: Nit: avoid making unrelated whitespace changes.

**Review by afd — APPROVED** [review]
Approved with one nit about an unnecessary whitespace change. Please revert that, and then we can run the CI bots and merge.

File: test/fuzz/transformation_access_chain_test.cpp (hunk 4 above)
Comment (sliu-UIUC) [reply test/fuzz/transformation_access_chain_test.cpp]: Done.

File: source/fuzz/transformation_access_chain.cpp [hunk 5]
```diff
@@ -363,9 +363,12 @@ protobufs::Transformation TransformationAccessChain::ToMessage() const {
   return result;
 }
 
-std::pair<bool, uint32_t> TransformationAccessChain::GetIndexValue(
+std::pair<bool, uint32_t> TransformationAccessChain::GetStructIndexValue(
     opt::IRContext* ir_context, uint32_t index_id,
     uint32_t object_type_id) const {
+  assert(ir_context->get_def_use_mgr()->GetDef(object_type_id) ==
+             SpvOpTypeStruct &&
```
Comment (afd) [inline source/fuzz/transformation_access_chain.cpp:370]: Oops: there's a problem here, which is that `GetDef()` returns an `Instruction*`, but what we want is the *opcode* of the instruction, so we need another `->opcode()`. Please check that this compiles without warnings and then commit. (The CI treats warnings as errors.)

**Review by afd — CHANGES_REQUESTED** [review]
The debug builds fail due to a comparison between a pointer and an enum value.

File: source/fuzz/transformation_access_chain.cpp (hunk 5 above)
Comment (sliu-UIUC) [reply source/fuzz/transformation_access_chain.cpp]: Solved.

**Comment by sliu-UIUC** [comment]
``test/fuzz/fuzzerutil_test.cpp`` can be ignored for this PR as it is implemented in another MaybeTest PR. Sorry for including it.

File: source/fuzz/transformation_access_chain.cpp [hunk 6]
```diff
@@ -433,4 +433,4 @@ std::unordered_set<uint32_t> TransformationAccessChain::GetFreshIds() const {
 }
 
 }  // namespace fuzz
-}  // namespace spvtools
+}  // namespace spvtools
```
Comment (afd) [inline source/fuzz/transformation_access_chain.cpp:436]: Please always include a newline at the end of a file.

**Review by afd — CHANGES_REQUESTED** [review]
Thanks for the fix! The only thing is that you don't have a new line at the end of the file.

File: source/fuzz/transformation_access_chain.cpp (hunk 6 above)
Comment (sliu-UIUC) [reply source/fuzz/transformation_access_chain.cpp:436]: Done.

**Review by afd — APPROVED** [review]
LGTM!
